# cleaning our text
In this section we will clean the bill text that we scraped from the last section. We will do this in the following steps:
1. get our text (if we don't already have it loaded up) from either scraping or loading the file from the previous section.
2. inspect our text to identify elements that we want to clean
3. write loops to remove these elements from the text
4. learn about functions so we can write one to clean our text in an automatic way
5. keep improving the funtion to clean more and more elements

Then, at the end, we will look at other tools (ChatGPT and OpenRefine) to help with the cleaning process. 

In [5]:
# run the lines below to load up the text from the course website

import requests
source = requests.get('https://bit.ly/transgender_text')
text = source.content

In [6]:
text[:100]

b'<html><body><pre>\n[Congressional Bills 117th Congress]\n[From the U.S. Government Publishing Office]\n'

In [7]:
type(text)

bytes

In [8]:
text = text.decode('utf-8')

In [5]:
# alternatively, uncomment the bottom four lines to load it from your own space
# notice that the data is already in a string format.

# load = open('sample.txt')
# loaded_text = load.read()
# load.close()
# loaded_text[:100]

## inspecting our text
Remember slicing? Take some slices of the text to see what elements we want to clean. Come up with a list of things that we want to remove.

In [9]:
text[:1000]

'<html><body><pre>\n[Congressional Bills 117th Congress]\n[From the U.S. Government Publishing Office]\n[H.R. 1112 Introduced in House (IH)]\n\n&lt;DOC&gt;\n\n\n\n\n\n\n117th CONGRESS\n  1st Session\n                                H. R. 1112\n\n   To require a report on the military coup in Burma, and for other \n                               purposes.\n\n\n_______________________________________________________________________\n\n\n                    IN THE HOUSE OF REPRESENTATIVES\n\n                           February 18, 2021\n\n    Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. \n  Buchanan) introduced the following bill; which was referred to the \n                      Committee on Foreign Affairs\n\n_______________________________________________________________________\n\n                                 A BILL\n\n\n \n   To require a report on the military coup in Burma, and for other \n                               purposes.\n\n    Be it enacted by the S

In [10]:
text[3000:4000]

"due to the Burmese military's \n        actions directly threatens the democratic trajectory of Burma's \n        Parliament, and thereby the country;\n            (3) the will and determination of those duly-elected \n        Members of Parliament who are taking it upon themselves to \n        continue serving as representatives of the people through \n        alternative methods of communicating and convening should be \n        lauded; and\n            (4) by preventing the Parliament from completing its work, \n        the Burmese military has rendered impossible and effectively \n        nullified the international collaborative relationships that \n        have supported and strengthened the institution, including the \n        Burmese parliament's partnership with HDP.\n\nSEC. 4. STATEMENT OF POLICY.\n\n    It is the policy of the United States to--\n            (1) engage with the Association of Southeast Asian Nations \n        (ASEAN) and ASEAN member states to--\n          

## looping through the text to `replace()` it
These are the elements we want to clean, as well as the large blank spaces: 

```
\n
/n
\\n
_
[
]
<html><body><pre>
<html><body><pre>
```

When you have a lot of items to remove at once, it's best to put them into a list. Then we can write a loop that goes through each item in the "take out" list to see if it's in the text data. If it is, we will replace that item with a blank space.

In [8]:
to_take_out = ['\n', '/n', '\\n', '_', '[', ']', '<html><body><pre>', '<html><body><pre>', '  ']

In [9]:
for item in to_take_out:
    if item in text:
        # here is a complicated line of code: 
        # we are replacing the item with nothing, indicated by two quotes 
        # then we are saving those results to "text", effectively overwriting
        # the variable. 
        text = text.replace(item, '')

In [10]:
text[:1000]

"Congressional Bills 117th CongressFrom the U.S. Government Publishing OfficeH.R. 1112 Introduced in House (IH)&lt;DOC&gt;117th CONGRESS1st SessionH. R. 1112 To require a report on the military coup in Burma, and for otherpurposes.IN THE HOUSE OF REPRESENTATIVES February 18, 2021Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs A BILLTo require a report on the military coup in Burma, and for otherpurposes.Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled,SECTION 1. SHORT TITLE.This Act may be cited as the ``Protect Democracy in Burma Act of 2021''.SEC. 2. FINDINGS.Congress finds the following:(1) On March 14, 2005, the House of Representatives agreed to H. Res. 135, which established the House Democracy Assistance Commission (later changed to the House Democracy Partnership, hereafter referred to as ``HDP'') to work d

## challenge: replacing with spaces

Notice that now some of the words in our dataset are joined with other words. That's because we took out the new lines. How could we change the code so that we keep spaces between words? 

## function to automate cleaning
Let's say we want to do this to many bits of text, not just one. We could automate the work by writing a function that can run on as many texts as we want. 

Functions have two key components: the definition and the call. You first define the function and what it does, then you "call" it to get it to work on a particular piece of data. 

Let's start with the definition. First, you name the function, and include parentheses for your parameters (more on this in a moment). Then, in the body of the definition, you write whatever python code you want to execute for that function. Finally, you have a return statement that saves or "returns" the result, so to speak, from the function. 

In [11]:
def add(x,y):
    answer = x + y
    return answer

Then we call the function.

In [12]:
add(5, 10021)

10026

The basic idea is that the input data, whatever data you want the function to work with, goes inside the parentheses. So that within the body of the function definition, that input data (known formally as "parameters") gets assigned to whatever variable is in the definition. 

This makes functions portable, so to speak, as you can write one, then call it using whatever input data that you like.

What would a function for our text cleaner look like?

In [13]:
# remove all the characters in the "take out" list by writing a
# loop that replaces those characters with an empty character, ''
def clean_up(data):
    to_take_out = ['\n', '/n', '\\n', '_', '[', ']', '<html><body><pre>', '<html><body><pre>', '  ']
    for item in to_take_out:
        if item in data:
            data = data.replace(item, '')
    return data

In [14]:
cleaned = clean_up(text)

In [15]:
cleaned[:1000]

"Congressional Bills 117th CongressFrom the U.S. Government Publishing OfficeH.R. 1112 Introduced in House (IH)&lt;DOC&gt;117th CONGRESS1st SessionH. R. 1112 To require a report on the military coup in Burma, and for otherpurposes.IN THE HOUSE OF REPRESENTATIVES February 18, 2021Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs A BILLTo require a report on the military coup in Burma, and for otherpurposes.Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled,SECTION 1. SHORT TITLE.This Act may be cited as the ``Protect Democracy in Burma Act of 2021''.SEC. 2. FINDINGS.Congress finds the following:(1) On March 14, 2005, the House of Representatives agreed to H. Res. 135, which established the House Democracy Assistance Commission (later changed to the House Democracy Partnership, hereafter referred to as ``HDP'') to work d

Finally, we save our text. And that's it!

In [34]:
with open('clean_sample.txt', 'w') as f:
    f.write(cleaned)

## cleaning with chatgpt
### prompt: clean all of the blank spaces and extraneous characters from this text with python.

In [11]:
import re
from html import unescape

def clean_text(text):
    # Remove HTML tags
    text = re.sub(r'<.*?>', '', text)
    
    # Unescape HTML entities
    text = unescape(text)
    
    # Remove extra blank lines and whitespace
    text = re.sub(r'\n\s*\n', '\n', text)  # Remove multiple newlines
    text = re.sub(r'\s+', ' ', text).strip()  # Normalize spaces

    return text

In [21]:
text = clean_text(text)

In [23]:
text[:10000]

"[Congressional Bills 117th Congress] [From the U.S. Government Publishing Office] [H.R. 1112 Introduced in House (IH)] <DOC> 117th CONGRESS 1st Session H. R. 1112 To require a report on the military coup in Burma, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES February 18, 2021 Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs _______________________________________________________________________ A BILL To require a report on the military coup in Burma, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Protect Democracy in Burma Act of 2021''. SEC. 2. FINDINGS. Congress finds the following: (1) On March 14, 2005, the House of Representatives agreed to H. Res

### follow-up prompt:
This text is a collection of legislative bills. Now write me some code that separates the long string of text into individual bills. 

In [24]:
def split_bills(text):
    # Define a regex pattern to detect the start of a new bill
    bill_pattern = re.compile(r'(?=\d{3}th CONGRESS\s+\d{1,2}st Session\s+H\. R\. \d+)', re.MULTILINE)
    
    # Split the text using the pattern
    bills = bill_pattern.split(text)
    
    # Remove empty entries and strip whitespace
    bills = [bill.strip() for bill in bills if bill.strip()]
    
    return bills


In [25]:
bills = split_bills(text)

In [26]:
type(bills)

list

In [27]:
len(bills)

11

In [29]:
bills[0]

'[Congressional Bills 117th Congress] [From the U.S. Government Publishing Office] [H.R. 1112 Introduced in House (IH)] <DOC>'

In [30]:
bills[1]

"117th CONGRESS 1st Session H. R. 1112 To require a report on the military coup in Burma, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES February 18, 2021 Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs _______________________________________________________________________ A BILL To require a report on the military coup in Burma, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Protect Democracy in Burma Act of 2021''. SEC. 2. FINDINGS. Congress finds the following: (1) On March 14, 2005, the House of Representatives agreed to H. Res. 135, which established the House Democracy Assistance Commission (later changed to the House Democracy Partnership, hereaft

In [31]:
bills[2]

"117th CONGRESS 1st Session H. R. 435 To exclude from tax certain payments of Federal pandemic unemployment compensation, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES January 21, 2021 Ms. Velazquez (for herself, Ms. Norton, Ms. Williams of Georgia, Mr. Smith of Washington, Ms. Jayapal, Mr. Meeks, Mr. Welch, Ms. Schakowsky, Ms. Kaptur, Ms. Tlaib, Ms. Wild, Ms. Newman, Ms. Lee of California, Ms. Meng, and Mr. Nadler) introduced the following bill; which was referred to the Committee on Ways and Means _______________________________________________________________________ A BILL To exclude from tax certain payments of Federal pandemic unemployment compensation, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Excluding Pandemic Unemployment Compensation from Income Ac

In [34]:
bills[5]

"117th CONGRESS 1st Session H. R. 269 To enhance the safety of ports of entry in the United States, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES January 11, 2021 Mr. Vela introduced the following bill; which was referred to the Committee on Transportation and Infrastructure _______________________________________________________________________ A BILL To enhance the safety of ports of entry in the United States, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Putting Our Resources Toward Security (PORTS) Act''. SEC. 2. IMPROVING PORTS OF ENTRY FOR BORDER SECURITY AND OTHER PURPOSES. (1) In general.--There are authorized to be appropriated to the Administrator of the General Services Administration $2,000,000,000 for each of fiscal years 2022 through 2024 to make i

In [35]:
bills[-1]

"117th CONGRESS 1st Session H. R. 154 To provide regulatory relief for Black and community banks, to codify the Minority Bank Deposit Program, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES January 4, 2021 Mr. Rush (for himself, Mr. Carson, Ms. Clarke of New York, Mr. Cooper, Mr. Khanna, and Mr. Welch) introduced the following bill; which was referred to the Committee on Financial Services, and in addition to the Committee on Ways and Means, for a period to be subsequently determined by the Speaker, in each case for consideration of such provisions as fall within the jurisdiction of the committee concerned _______________________________________________________________________ A BILL To provide regulatory relief for Black and community banks, to codify the Minority Bank Deposit Program, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congr

### follow-up prompt:
You split the first bill incorrectly. make sure the first bill header information is included with the body of the bill, like on the other bills

In [38]:
def split_bills(text):
    # Regular expression to detect the start of a new bill
    bill_start_pattern = re.compile(r"\n\[Congressional Bills \d+th Congress\]\n.*?\n.*?Introduced in House \(IH\)\]\n")
    
    # Find all start positions of bills
    matches = list(bill_start_pattern.finditer(text))
    
    bills = []
    for i in range(len(matches)):
        start = matches[i].start()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        bills.append(text[start:end].strip())
    
    return bills

In [41]:
bills = split_bills(text)

In [42]:
bills

[]

### follow-up prompt
Now I'm getting nothing for bills

In [43]:
import re

def split_bills(text):
    # Regular expression to detect the start of a new bill
    bill_start_pattern = re.compile(r"(\[Congressional Bills \d+th Congress\].*?Introduced in House \(IH\)\])", re.DOTALL)

    # Split based on detected bill headers
    parts = bill_start_pattern.split(text)
    
    # Reconstruct bills ensuring the header stays with the content
    bills = []
    for i in range(1, len(parts), 2):
        header = parts[i].strip()
        body = parts[i + 1].strip() if i + 1 < len(parts) else ""
        bills.append(f"{header}\n\n{body}")

    return bills

bills = split_bills(text)


In [45]:
bills[0]

"[Congressional Bills 117th Congress] [From the U.S. Government Publishing Office] [H.R. 1112 Introduced in House (IH)]\n\n<DOC> 117th CONGRESS 1st Session H. R. 1112 To require a report on the military coup in Burma, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES February 18, 2021 Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs _______________________________________________________________________ A BILL To require a report on the military coup in Burma, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Protect Democracy in Burma Act of 2021''. SEC. 2. FINDINGS. Congress finds the following: (1) On March 14, 2005, the House of Representatives agreed to H. 

### follow-up prompt
Extract metadata for title, number, session, and plain text

In [46]:

def extract_metadata(bill_text):
    metadata = {}
    
    # Extract bill number
    bill_number_match = re.search(r"\[Congressional Bills (\d+th Congress)\]", bill_text)
    metadata['session_number'] = bill_number_match.group(1) if bill_number_match else "Unknown"
    
    # Extract title
    title_match = re.search(r"\n(.*?)\n.*?Introduced in House", bill_text)
    metadata['title'] = title_match.group(1).strip() if title_match else "Unknown"
    
    # Extract bill number from text header
    bill_id_match = re.search(r"(H\.R\. \d+|S\. \d+)", bill_text)
    metadata['bill_number'] = bill_id_match.group(1) if bill_id_match else "Unknown"
    
    # Store the plain text content
    metadata['text'] = bill_text.strip()
    
    return metadata

In [48]:
extract_metadata(bills[0])

{'session_number': '117th Congress',
 'title': 'Unknown',
 'bill_number': 'H.R. 1112',
 'text': "[Congressional Bills 117th Congress] [From the U.S. Government Publishing Office] [H.R. 1112 Introduced in House (IH)]\n\n<DOC> 117th CONGRESS 1st Session H. R. 1112 To require a report on the military coup in Burma, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES February 18, 2021 Mr. Connolly (for himself, Mr. Price of North Carolina, and Mr. Buchanan) introduced the following bill; which was referred to the Committee on Foreign Affairs _______________________________________________________________________ A BILL To require a report on the military coup in Burma, and for other purposes. Be it enacted by the Senate and House of Representatives of the United States of America in Congress assembled, SECTION 1. SHORT TITLE. This Act may be cited as the ``Protect Democracy in Burma Act of 2021''. SEC. 2. FINDINGS.

In [49]:
extract_metadata(bills[-1])

{'session_number': '117th Congress',
 'title': 'Unknown',
 'bill_number': 'H.R. 154',
 'text': "[Congressional Bills 117th Congress] [From the U.S. Government Publishing Office] [H.R. 154 Introduced in House (IH)]\n\n<DOC> 117th CONGRESS 1st Session H. R. 154 To provide regulatory relief for Black and community banks, to codify the Minority Bank Deposit Program, and for other purposes. _______________________________________________________________________ IN THE HOUSE OF REPRESENTATIVES January 4, 2021 Mr. Rush (for himself, Mr. Carson, Ms. Clarke of New York, Mr. Cooper, Mr. Khanna, and Mr. Welch) introduced the following bill; which was referred to the Committee on Financial Services, and in addition to the Committee on Ways and Means, for a period to be subsequently determined by the Speaker, in each case for consideration of such provisions as fall within the jurisdiction of the committee concerned _______________________________________________________________________ A BILL To p

In [50]:
bills_metadata = [extract_metadata(bill) for bill in bills]

In [53]:
type(bills_metadata)

list

In [55]:
import pandas as pd

df = pd.DataFrame(bills_metadata)

In [56]:
df

session_number    title bill_number  \
0  117th Congress  Unknown   H.R. 1112   
1  117th Congress  Unknown    H.R. 435   
2  117th Congress  Unknown    H.R. 886   
3  117th Congress  Unknown    H.R. 464   
4  117th Congress  Unknown    H.R. 269   
5  117th Congress  Unknown    H.R. 144   
6  117th Congress  Unknown    H.R. 902   
7  117th Congress  Unknown   H.R. 4960   
8  117th Congress  Unknown     H.R. 32   
9  117th Congress  Unknown    H.R. 154   

                                                text  
0  [Congressional Bills 117th Congress] [From the...  
1  [Congressional Bills 117th Congress] [From the...  
2  [Congressional Bills 117th Congress] [From the...  
3  [Congressional Bills 117th Congress] [From the...  
4  [Congressional Bills 117th Congress] [From the...  
5  [Congressional Bills 117th Congress] [From the...  
6  [Congressional Bills 117th Congress] [From the...  
7  [Congressional Bills 117th Congress] [From the...  
8  [Congressional Bills 117th Congress] [From the...  
9  [Congressional Bills 117th Congress] [From the...

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   session_number  10 non-null     object
 1   title           10 non-null     object
 2   bill_number     10 non-null     object
 3   text            10 non-null     object
dtypes: object(4)
memory usage: 452.0+ bytes


In [58]:
df['text']

0    [Congressional Bills 117th Congress] [From the...
1    [Congressional Bills 117th Congress] [From the...
2    [Congressional Bills 117th Congress] [From the...
3    [Congressional Bills 117th Congress] [From the...
4    [Congressional Bills 117th Congress] [From the...
5    [Congressional Bills 117th Congress] [From the...
6    [Congressional Bills 117th Congress] [From the...
7    [Congressional Bills 117th Congress] [From the...
8    [Congressional Bills 117th Congress] [From the...
9    [Congressional Bills 117th Congress] [From the...
Name: text, dtype: object

In [104]:
# Function to extract metadata from a bill
def extract_metadata(bill_text):
    bill_number_match = re.search(r"H\.R\. \d+", bill_text)
    
    # Find the third number occurrence (bill number and other numbers)
    number_matches = re.findall(r"\d+", bill_text)
    if len(number_matches) >= 3:
        # Find text after the third occurrence of numbers, stopping at the first underscore
        title_start = bill_text.find(number_matches[2]) + len(number_matches[2])  # After the third number
        title_end = bill_text.find("_", title_start)  # Stop at the first underscore
        title = bill_text[title_start:title_end].strip() if title_end != -1 else bill_text[title_start:].strip()
    else:
        title = "No Title Found"
    
    session_match = re.search(r"\[Congressional Bills (\d+th Congress)\]", bill_text)
    
    bill_number = bill_number_match.group(0) if bill_number_match else "Unknown"
    session = session_match.group(1) if session_match else "Unknown Session"
    
    return {
        "Bill Number": bill_number,
        "Title": title,
        "Session": session,
        "Full Text": bill_text
    }


In [105]:
bills_metadata = [extract_metadata(bill) for bill in bills]

In [106]:
df = pd.DataFrame(bills_metadata)

In [107]:
df

Bill Number                                              Title  \
0   H.R. 1112  th Congress] [From the U.S. Government Publish...   
1    H.R. 435  th Congress] [From the U.S. Government Publish...   
2    H.R. 886  th Congress] [From the U.S. Government Publish...   
3    H.R. 464  th Congress] [From the U.S. Government Publish...   
4    H.R. 269  th Congress] [From the U.S. Government Publish...   
5    H.R. 144  th Congress] [From the U.S. Government Publish...   
6    H.R. 902  th Congress] [From the U.S. Government Publish...   
7   H.R. 4960  th Congress] [From the U.S. Government Publish...   
8     H.R. 32  th Congress] [From the U.S. Government Publish...   
9    H.R. 154  th Congress] [From the U.S. Government Publish...   

          Session                                          Full Text  
0  117th Congress  [Congressional Bills 117th Congress] [From the...  
1  117th Congress  [Congressional Bills 117th Congress] [From the...  
2  117th Congress  [Congressional Bills 117th Congress] [From the...  
3  117th Congress  [Congressional Bills 117th Congress] [From the...  
4  117th Congress  [Congressional Bills 117th Congress] [From the...  
5  117th Congress  [Congressional Bills 117th Congress] [From the...  
6  117th Congress  [Congressional Bills 117th Congress] [From the...  
7  117th Congress  [Congressional Bills 117th Congress] [From the...  
8  117th Congress  [Congressional Bills 117th Congress] [From the...  
9  117th Congress  [Congressional Bills 117th Congress] [From the...

### follow-up prompt
To get the title, get the text immediately following the third group of consecutive numbers, up until the first underscore

In [100]:
# Function to extract metadata from a bill
def extract_metadata(bill_text):
    bill_number_match = re.search(r"H\.R\. \d+", bill_text)
    
    # Find all groups of consecutive numbers
    number_matches = re.findall(r"\d+", bill_text)
    
    if len(number_matches) >= 3:
        # Find the third group of consecutive numbers and extract text following it
        third_number_position = bill_text.find(number_matches[2]) + len(number_matches[2])  # After the third number
        title_start = third_number_position
        
        # Stop at the first underscore
        title_end = bill_text.find("_", title_start)
        
        # If underscore is found, take text until the underscore, otherwise take all remaining text
        title = bill_text[title_start:title_end].strip() if title_end != -1 else bill_text[title_start:].strip()
    else:
        title = "No Title Found"
    
    # Extract session (if present)
    session_match = re.search(r"\[Congressional Bills (\d+th Congress)\]", bill_text)
    
    bill_number = bill_number_match.group(0) if bill_number_match else "Unknown"
    session = session_match.group(1) if session_match else "Unknown Session"
    

In [101]:
bills_metadata = [extract_metadata(bill) for bill in bills]

In [102]:
df = pd.DataFrame(bills_metadata)

In [103]:
df

0
0  None
1  None
2  None
3  None
4  None
5  None
6  None
7  None
8  None
9  None

In [108]:
df.to_csv('trans.csv')

## cleaning tabular data with OpenRefine

[OpenRefine](https://openrefine.org/) is an open source GUI tool for working with tabular data. It is especially useful for cleaning and organizing spreadsheet data. 

We are going to use this tool to clean the "Title" column in our `trans.csv` dataset.

First, download and install the app. [Download here](https://openrefine.org/download).

After installing, when you open the app, it will appear on your browser, like Jupyter Notebooks. 

Second, we will import our dataset. On the application page, click the "browese" button, select your data source, and press "next". On the following page, you'll see a bunch of options for customizing your project. We don't have to change anything, so press "Create Project" on the right. 

There is so much you can to with OpenRefine, particularly with [Facets](https://openrefine.org/docs/manual/facets) and [Filters](https://openrefine.org/docs/manual/facets#text-filter), which allow you to edit the spreadsheet by applying changes to patterns across cells rather than individual cells. Let's say you want to take out some unwanted html elements that appear in your text, or you want to regularize the capitalization or spelling of a certain word or phrase. With facets and filtering, you can easily find the pattern and fix it. See more on this [tutorial video](https://www.youtube.com/watch?v=B70J_H_zAWM&t=73s).

Finally, we will do a bit of cleaning.

### cleaning the "title" column
In our case, we are going to work on just one column, the "Title" column. We want to delete the text that comes before the title. 

Go to the column that says "Title," open the dropdown menu, hover over "edit column" and select "split into several columns"

![Split into several columns option](split.png)

Then, sure the "by separator" option is selected, which it should be by default. Then, you will choose your "separator," in the text box below. A separator is the element by which OpenRefine will know where to split the column. For our dataset, we can use the phrase "Session H.R. 1112" to split our columns. That way, we can get just the title in its own column, and get rid of the other (original) column.

Now, we need to generate some regex. For this, I went back to Chat GPT and asked it to generate some regex that finds the pattern. I had to combine these results with what I found on the [regex generator](https://regex-generator.olafneumann.org/), in order to get the code to work. 

![chatgpt regex](chat.png)

Using that, I generated the string, `Session\s[A-Za-z]\.\s[A-Za-z]\.\s\d{2,4}`, and input it into the split cells window, making sure to select the regex tickbox. 

![splitting cells by regex expression](split2.png)

After this, I got rid of the column, renamed the column I did keep to "Title," and did a "transform" (selecting "common transform") of trailing whitespace. All of these are under the options "edit column" on the dropdown.

And that's it! Now my dataset is much cleaner. 

![title column cleaned up](cleaned.png)
